### ISSUE TITLE
Intermittent two-part/two-pass range endpoint off-by-one (related to 'spesh delay'?)

### ISSUE THREAD

**Issue by SqrtNegInf** [rakudo/rakudo#2805]
## The Problem
There is an off-by-one for upper end of range `$x ^..^ @y` (counts one past end of array) when used in the context of a 'for' loop.  The bug is resistant to golfing, and–to make things even harder–doesn't trigger with each run.

## Steps to Reproduce
Given the code below, the bug happens when 'anastates' is called a 2nd (or 3rd) time if both 'trigger A' and 'trigger B' are active.

There are two ways to avoid the bug:
• if 'trigger A' or 'trigger B' is disabled (and a corresponding 'work-around' A/B is used)
• if MVM_SPESH_NODELAY=1 is set

```
my @states = < 0 1 2 3 4 5 6 7 8 9 A B C D E F G H I K L M N O P Q R S T U V W X Y Z a b c d e f >;
say 'RUN 1'; .say for my @res1 = anastates @states;
say 'RUN 2'; .say for my @res2 = anastates @states; # problem is always on 2nd (or 3rd) call

sub anastates (@s) {
    my @pairs;
    for ^@s -> $i {
       #for $i ^..  @s.end -> $j {  # work-around A1
       #for $i ^..^ +@s -> $j {     # work-around A2
       #my $range = $i ^..^ @s;     # work-around A3.1
       #for $range -> $j {          # work-around A3.2
        for $i ^..^ @s -> $j {      # trigger A
            die "BUG" if ! defined @s[$j]; # just bail when $j doesn't subscript a valid location in @s
            @pairs.push: [ @s[$i], @s[$j] ];
        }
    }
    my $equivs = hash @pairs.classify: *.lc.comb.sort.join; #  warnings when @pairs contains 'Any')
    gather for $equivs.values -> @c {
    for ^@c -> $i {
       #for $i ^.. @c.end {         # work-around B1
       #for $i ^..^ +@c {           # work-around B2
        for $i ^..^ @c {            # trigger B
            # this code not relevant
        }
    }
    }
}
```

Even more information here:
https://github.com/SqrtNegInf/Rosettacode-Perl6-Smoke/blob/master/S/State_name_puzzle-debug

## Environment
First seen approximately 2019-03-15:
Rakudo version 2019.03-39-g23fca8f6f built on MoarVM version 2019.03-31-g6c7810ce7

**Comment by AlexDaniel** [comment]
Very nice find. But… this is a regression, right? I can only reproduce it on HEAD.

**Comment by AlexDaniel** [comment]
Yeah, I definitely cannot trigger it on any release: https://gist.github.com/4e4c7bc46e1c37e0a826bf4a2b9a8837

So with that snippet I was able to bisect it. Offending commits are:
* 7b863b72aac89ddee02826266e28685b12b1f74f
* f09156240a52b34bc356ef715bd955c8f4378cb4
* 7bd3b3a0a8a3c1faaef602d2d3fe1e619e4ef640

I don't know if these actually introduce an issue, or if these simply expose it in some way. In any case, ping @jnthn.

**Comment by AlexDaniel** [comment]
Ah, actually, there are many commits in that last bump. See:
* https://github.com/perl6/nqp/compare/2019.03-7-g87771ea33...2019.03-19-gda96d39
* https://github.com/MoarVM/MoarVM/compare/2019.03-2-gbb2dc6619...2019.03-31-g6c7810c

So if someone feels like it, you're welcome to bisect moar itself :)

**Comment by dogbert17** [comment]
There's a decent chance that the offending commit is https://github.com/MoarVM/MoarVM/commit/bc98cb9467ab2d5d408dff841f39ae8734a936a5

**Comment by SqrtNegInf** [comment]
Very encouraging, preliminary results look good... Will continue testing to ensure this is solely responsible for the problem, but you definitely identified the crucial part.

**Comment by SqrtNegInf** [comment]
Yes, the blame for this regression falls 100% on [MoarVM/MoarVM@bc98cb9]

**Comment by jnthn** [comment]
I've tried to reproduce this on HEAD builds of MoarVM and Rakudo today, and failed to do so no matter what spesh flags I set, in dozens of runs. Wonder if it's been fixed by one of the other changes in the meantime...

**Comment by dogbert17** [comment]
I'm also unable to reproduce on HEAD. A bisect points to the problem being fixed (masked ?) with commit a21cf6ae0b36ca69fad2db591e6d7107b4a6ba04

**Comment by SqrtNegInf** [comment]
Agreed, the problem appears to be gone. Easy come, easy go...
